orttraining/orttraining/training_api/checkpoint.cc

```diff
@@ -1,581 +1,720 @@
+      optimizer_states.emplace_back(fbs_optimizer_state);
     }
 
-    // Save each optimizer state (of all parameters) into single file.
-    // For example: save "momentum_1" of all parameters into one file.
-    for (auto& pair : optimizer_state_ort_values) {
-      const PathString momentum_name = ToPathString(pair.first);
-      const std::unordered_map<std::string, OrtValue>& param_name_to_ortvalue = pair.second;
-      const PathString& cur_state_filename_prefix =
-          StringConcat(cur_group_filename_prefix, momentum_name);
-
-      std::vector<ONNX_NAMESPACE::TensorProto> saved_tensor_protos;
-      ORT_RETURN_IF_ERROR(CreateTensorProtosFromOrtValues(
-          param_name_to_ortvalue,
-          *optimizer_state.optimizer_session_data_transfer_mgr,
-          saved_tensor_protos));
-
-      // Save TensorProto to file.
-      WriteTensorProtoToFile(
-          GetTensorProtoFilePath(checkpoint_path, cur_state_filename_prefix),
-          saved_tensor_protos, "[optimizer_state]");
-    }
+    const auto fbs_group_name = builder.CreateString(group_named_optimizer_state.first);
+    const auto fbs_optimizer_states = builder.CreateVector(optimizer_states);
+
+    fbs::OptimizerGroupBuilder optimizer_state_builder(builder);
+    optimizer_state_builder.add_group_name(fbs_group_name);
+    optimizer_state_builder.add_initial_learning_rate(group_optimizer_state_ptr->initial_lr);
+    optimizer_state_builder.add_step(group_optimizer_state_ptr->step);
+    optimizer_state_builder.add_optimizer_states(fbs_optimizer_states);
+    auto fbs_optimizer_group = optimizer_state_builder.Finish();
+    fbs_optimizer_groups.emplace_back(fbs_optimizer_group);
+  }
+
+  return Status::OK();
+}
 
-    // Storing group-wise properties.
-    PropertyBag properties;
-    properties.AddProperty(builtin_lr_property_name, group_optimizer_state_ptr->initial_lr);
-    properties.AddProperty(builtin_step_property_name, group_optimizer_state_ptr->step);
-    std::vector<ONNX_NAMESPACE::TensorProto> group_wise_properties_tensor_protos;
-    properties.ToTensorProtos(group_wise_properties_tensor_protos);
+/**
+ * @brief Save from user defined properties to a flatbuffer checkpoint property bag.
+ *
+ * @param property_bag user defined properties.
+ * @param builder Flatbuffer builder.
+ * @param fbs_property_bag Flatbuffer property bag to be populated.
+ * @return Status of the operation.
+ */
+Status FromPropertyBag(const PropertyBag& property_bag, flatbuffers::FlatBufferBuilder& builder,
+                       flatbuffers::Offset<fbs::PropertyBag>& fbs_property_bag) {
+  if (property_bag.Size() == 0U) {
+    return Status::OK();
+  }
 
-    WriteTensorProtoToFile(
-        GetTensorProtoPropertiesFilePath(checkpoint_path, cur_group_filename_prefix),
-        group_wise_properties_tensor_protos, "[param_group_properties]");
+  std::vector<flatbuffers::Offset<fbs::IntProperty>> ints;
```

Comment:
similar comment about sorting keys for property bag before writing to flatbuffers format. there may be other unordered containers I've missed too